### PR TITLE
Fixes #17, code refactoring, and doc polish

### LINKED ISSUE #17: `Error` type should implement `std::error::Error`.
Common error chaining crates (`error-chain`, `failure`, etc...) require all error types in their chains implement the trait `std::error::Error`. Currently, `reed_solomon_erasure::Error` does not implement this trait, thus excluding it from being used alongside many popular approaches to error handling.

### PR DESCRIPTION
**Pull request by darrenldl**
- Fixes #17 
- Some code refactoring and optimization
- Doc polish

### CODE REVIEW

**Comment by darrenldl** [comment]
@DrPeterVanNostrand
I am not really sure what to put in the description strings, would appreciate some feedback on the added strings.

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name    = "reed-solomon-erasure"
-version = "3.0.5"
+version = "3.1.0"
```
Comment (darrenldl) [inline Cargo.toml]: Version bump, no code changes

File: build.rs [hunk 2]
```diff
@@ -46,8 +46,7 @@ fn multiply(log_table : &[u8; FIELD_SIZE],
             b : u8) -> u8 {
     if a == 0 || b == 0 {
         0
-    }
-    else {
+    } else {
```
Comment (darrenldl) [inline build.rs]: Simple formatting fix, nothing problematic

File: src/galois.rs [hunk 3]
```diff
@@ -16,13 +16,11 @@ pub fn mul(a : u8, b : u8) -> u8 {
 }
 
 pub fn div(a : u8, b : u8) -> u8 {
-    if a == 0 {
+    if        a == 0 {
```
Comment (darrenldl) [inline src/galois.rs]: Simple spacing fix, nothing problematic needs reviewing

File: src/galois.rs [hunk 4]
```diff
@@ -16,13 +16,11 @@ pub fn mul(a : u8, b : u8) -> u8 {
...
         0
-    }
-    else if b == 0 {
+    } else if b == 0 {
```
Comment (darrenldl) [inline src/galois.rs]: Simple formatting fix, nothing problematic needs reviewing

File: src/lib.rs [hunk 5]
```diff
@@ -1385,7 +1368,7 @@ impl ReedSolomon {
                 }
             }
             {
-                // Gather up the references of all data slices
+                // Gather up all the data slices
```
Comment (darrenldl) [inline src/lib.rs]: Dev comments update, nothing problematic

**Comment by DrPeterVanNostrand** [comment]
I think that using excerpts from the docs for `struct ReedSoloman` would provide adequate context for most of the error descriptions. For example (in `errors.rs`), maybe something like:

```
impl Error {
    fn to_string(&self) -> &str {
        match *self {
            Error::TooFewShards => "The number of provided shards is fewer than the number of shards in the codec",
            Error::TooManyShards => "The number of provided shards is more than the number of shards in the codec",
            Error::TooFewDataShards => "The number of data shards is less than the number of shards in the codec",
            ... ,
            Error::IncorrectShardSize => "The provided shards are of differing lengths",
            Error::EmptyShard => "The first provided shard has a length of zero",
            Error::InvalidShardFlags => "The number of flags does not match the total number of shards",
            Error::InvalidIndex => "Invalid index"
        }
    }
}

impl std::error::Error for Error {
    fn description(&self) -> &str {
        self.to_string()
    }
}
```

**Comment by DrPeterVanNostrand** [comment]
Yup, those error descriptions look good to me. Thanks again for getting this change in so quickly, it is much appreciated!

**Pull request merged**